src/platform/ESP32/OTAImageProcessorImpl.cpp

```diff
@@ -0,0 +1,210 @@
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+    imageProcessor->mOTAUpdatePartition = esp_ota_get_next_update_partition(NULL);
+    if (imageProcessor->mOTAUpdatePartition == NULL)
+    {
+        ChipLogError(SoftwareUpdate, "OTA partition not found");
+        return;
+    }
+    esp_err_t err =
+        esp_ota_begin(imageProcessor->mOTAUpdatePartition, OTA_WITH_SEQUENTIAL_WRITES, &(imageProcessor->mOTAUpdateHandle));
+    if (err != ESP_OK)
+    {
+        imageProcessor->mDownloader->OnPreparedForDownload(ESP32Utils::MapError(err));
+        return;
+    }
+    imageProcessor->mDownloader->OnPreparedForDownload(CHIP_NO_ERROR);
+}
+
+void OTAImageProcessorImpl::HandleFinalize(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    esp_err_t err = esp_ota_end(imageProcessor->mOTAUpdateHandle);
+    if (err != ESP_OK)
+    {
+        if (err == ESP_ERR_OTA_VALIDATE_FAILED)
+        {
+            ESP_LOGE(TAG, "Image validation failed, image is corrupted");
+        }
+        else
+        {
+            ESP_LOGE(TAG, "esp_ota_end failed (%s)!", esp_err_to_name(err));
+        }
+        return;
+    }
+    imageProcessor->ReleaseBlock();
+    ChipLogProgress(SoftwareUpdate, "OTA image downloaded to offset 0x%x", imageProcessor->mOTAUpdatePartition->address);
+}
+
+void OTAImageProcessorImpl::HandleAbort(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    if (esp_ota_abort(imageProcessor->mOTAUpdateHandle) != ESP_OK)
+    {
+        ESP_LOGE(TAG, "ESP OTA abort failed");
+    }
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+    esp_err_t err = esp_ota_write(imageProcessor->mOTAUpdateHandle, imageProcessor->mBlock.data(), imageProcessor->mBlock.size());
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "esp_ota_write failed (%s)", esp_err_to_name(err));
+        imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
+    }
+    imageProcessor->mParams.downloadedBytes += imageProcessor->mBlock.size();
+    imageProcessor->mDownloader->FetchNextData();
+}
+
+void OTAImageProcessorImpl::HandleApply(intptr_t context)
```

Comment:
When will the device reset?

Reply:
The device will reset and apply the new image after you reboot it manually.

Reply:
Why do you have to reboot the device manually?

Reply:
@dhrishi Rebooting the device manually will make sure that the device will receive the response of `applyimage` from the provider. The device is allowed to reboot only after receiving the response.